### PR TITLE
Whitespace

### LINKED ISSUE #18: Better whitespace combinators that don't write immediately
We can do it with functions or an intermediate data type, so for example either:
 - reintroduce Preceding and make the whitespace combinators modify the state to keep track of what should be written, without writing it. The verbatim combinator will write the requested whitespace
 - or don't write ByteString chunks but an intermediate representation (ADT) which still has to be interpreted with the logic above

### PR DESCRIPTION
**Pull request by roberth**
Alternative approach of tracking whitespace via state wasn't feasible `trySingleLine`.

Closes #18

### CODE REVIEW

File: canonix/canonix.cabal [hunk 1]
```diff
@@ -19,8 +19,11 @@ common src-main
                      , comonad
                      , containers
                      , free
```
Comment (domenkozar) [inline canonix/canonix.cabal:21]: don't need free anymore

**Review by domenkozar — CHANGES_REQUESTED** [review]
It seems to erase all newlines is that expected?

File: canonix/canonix.cabal (hunk 1 above)
Comment (roberth) [reply canonix/canonix.cabal:21]: I thought the same, but we're still using `Cofree` for trees. :coffee:

**Comment by roberth** [comment]
Ignoring newlines already happened, but now we're generating fewer.
To preserve newlines, we can automatically track line numbers between `verbatim` calls and request more as appropriate.

File: canonix/canonix.cabal (hunk 1 above)
Comment (domenkozar) [reply canonix/canonix.cabal:21]: isn't that part of `comonad`?

Comment (roberth) [reply canonix/canonix.cabal:21]: One would think, but no. https://hackage.haskell.org/package/free-5.1/docs/Control-Comonad-Cofree.html#t:Cofree

**Review by domenkozar — APPROVED** [review]

**Pull request merged**